index.js

```diff
@@ -144,13 +144,136 @@ function hash(msgVal) {
+ * Performs message validation before returning the decoded message object.
+ *
+ * @param {Buffer} bbmsg a bendy-butt message encoded with `bencode`
+ * @returns {Object} an object compatible with ssb/classic `msg.value` or an `Error`
+ */
+function decodeAndValidateSingle(bbmsg, previousMsg, hmacKey) {
+  if (bbmsg.length > 8192)
+    return new Error('invalid message: length must not exceed 8192 bytes')
+
+  const msgBFE = bencode.decode(bbmsg)
+
+  if (!Array.isArray(msgBFE) || msgBFE.length !== 2)
+    return new Error('invalid message: must be a list of payload and signature')
+
+  const [payload, signature] = bfe.decode(msgBFE)
+
+  if (!Array.isArray(payload) || payload.length !== 5)
+    return new Error(
+      'invalid payload: must be a list of author, sequence, previous, timestamp and contentSection'
+    )
+
+  const typeFormatErr = validateTypeFormat(msgBFE)
+  if (typeFormatErr) return typeFormatErr
+
+  const [author, sequence, previous, timestamp, contentSection] = payload
+
+  const signatureErr = validateSignature(author, payload, signature, hmacKey)
+  if (signatureErr) return signatureErr
+
+  const previousErr = validatePrevious(author, sequence, previous, previousMsg)
+  if (previousErr) return previousErr
+
+  const msgVal = {
+    author,
+    sequence,
+    previous,
+    timestamp,
+    signature,
+  }
+  if (isBoxedString(contentSection)) {
+    msgVal.content = contentSection
+  } else {
+    const [content, contentSignature] = contentSection
+    msgVal.content = content
+    msgVal.contentSignature = contentSignature
+  }
+
+  return msgVal
+}
+
+function validateSignature(author, payload, signature, hmacKey) {
+  const hmacKeyErr = validateHmacKey(hmacKey)
+  if (hmacKeyErr) return hmacKeyErr
+
+  if (!ssbKeys.verify(author, signature, hmacKey, payload))
+    return new Error(
+      'invalid message: signature must correctly sign the payload'
+    )
+}
+
+function validatePrevious(author, sequence, previous, previousMsg) {
+  if (sequence === 1) {
+    if (previous !== null)
+      return new Error(
+        'invalid message: message must have a previous value of null if sequence is 1'
+      )
+  } else {
+    const previousMsgAuthor = previousMsg[0]
+    if (!previousMsg)
+      return new Error(
+        'invalid previousMsg: value must not be undefined if sequence > 1'
+      )
+    if (author !== previousMsgAuthor)
+      return new Error(
+        'invalid message: author must be the same for the current and previous messages'
+      )
+
+    const previousHash = hash(previousMsg)
+    if (previous !== previousHash)
+      return new Error(
+        'invalid message: expected different previous message on feed'
+      )
+  }
+}
+
+function validateTypeFormat(msgBFE) {
+  const payload = msgBFE[0]
+  const [author, sequence, previous] = payload
+
+  if (author.slice(0, 2).toString('hex') !== '0003')
+    return new Error(
+      'invalid message: author value must have the correct type-format'
```

Comment:
We could make these error messages a bit more useful for programmers who encounter it. Instead of stating the expected result ("must have the correct..."), we could state what was the actual result, because this speeds up the debugging process. E.g. if the error said `invalid message: author field 0x0002 is incorrect`, then the programmer knows that a `0002` TFD slipped through, and that probably already gives them some information on what went wrong. Without this information, they would be forced to patch an extra console.log to figure out what went wrong, and that's assuming they have the chance of adding console.log (say, for instance, that someone else sends us a bug report saying "invalid message: author value must have the correct type-format", we still don't know what to do to fix it).

So this suggestion applies to the majority of these error messages, let's try to make them display the faulty data to speed up debugging.

Reply:
Thanks for the kind words and terrific feedback; I could not have hoped for better mentors and team mates.

Great suggestion regarding error messages! I've added the actual result to each message (where appropriate) and included a brief description of the expected value; hopefully this offers the best of both worlds when debugging.